### PR TITLE
fix insecure random source for tokens

### LINKED ISSUE #183: [SECURITY] CWE-338: Vulnerability in JHipster Kotlin
Due to the uncontrolled nature of the previous full disclosure of this vulnerability in `jhipster-kotlin` this is an 0-Day vulnerability disclosure for jhipster-kotlin.

See: https://github.com/jhipster/generator-jhipster/issues/10401
https://github.com/jhipster/generator-jhipster/security/advisories/GHSA-mwp6-j9wf-968c

There is currently no patch available at this time.

---

# [CWE-338: Use of Cryptographically Weak Pseudo-Random Number Generator (PRNG)](https://cwe.mitre.org/data/definitions/338.html)

JHipster is using an insecure source of randomness to generate all of it's random values. JHipster relies upon apache commons lang3 `RandomStringUtils`.

From the documentation:

> Caveat: Instances of Random, upon which the implementation of this class relies, are not cryptographically secure.
> \- https://commons.apache.org/proper/commons-lang/javadocs/api-3.9/org/apache/commons/lang3/RandomStringUtils.html

Here are the examples of JHipster Kotlin's use of an insecure PRNG:

https://github.com/jhipster/jhipster-kotlin/blob/193ae8f13c0be686f9687e78bacfedb144c47d8c/generators/server/templates/src/main/kotlin/package/service/util/RandomUtil.kt.ejs#L27-L62

## Proof Of Concepts Already Exist

There has been a POC of taking one RNG value generated `RandomStringUtils` and reversing it to generate all of the past/future RNG values public since March 3rd, 2018.

https://medium.com/@alex91ar/the-java-soothsayer-a-practical-application-for-insecure-randomness-c67b0cd148cd

POC Repository: https://github.com/alex91ar/randomstringutils

## Potential Impact Technical

All that is required is to get one password reset token from a JHipster generated service and using the POC above, you can reverse what all future password reset tokens to be generated by this server. This allows an attacker to pick and choose what account they would like to takeover by sending account password reset requests for targeted accounts.

### PR DESCRIPTION
**Pull request by atomfrede**
closes #183

### CODE REVIEW

File: generators/server/templates/src/main/kotlin/package/service/util/RandomUtil.kt.ejs [hunk 1]
```diff
@@ -20,44 +20,51 @@
...
- * Generate a password.
- *
- * @return the generated password.
- */
-fun generatePassword(): String = RandomStringUtils.randomAlphanumeric(DEF_COUNT)
+* Generate a password.
+*
+* @return the generated password.
+*/
+fun generatePassword(): String = generateRandomAlphanumericString()
```
Comment (sendilkumarn) [inline generators/server/templates/src/main/kotlin/package/service/util/RandomUtil.kt.ejs]:
```suggestion
fun generatePassword() = generateRandomAlphanumericString()
```

File: generators/server/templates/src/main/kotlin/package/service/util/RandomUtil.kt.ejs [hunk 2]
```diff
@@ -20,44 +20,51 @@
...
 package <%=packageName%>.service.util
 
+import java.security.SecureRandom
 import org.apache.commons.lang3.RandomStringUtils
 
 private const val DEF_COUNT = 20
 
+private val secureRandom: SecureRandom = SecureRandom().apply{ nextBytes(ByteArray(64)) }
+
+private fun generateRandomAlphanumericString(): String {
```
Comment (sendilkumarn) [inline generators/server/templates/src/main/kotlin/package/service/util/RandomUtil.kt.ejs]:
```suggestion
private fun generateRandomAlphanumericString() =
```

**Comment by sendilkumarn** [comment]
Now the tests are failing because of unrelated changes. We will have to fix the build now.

**Comment by atomfrede** [comment]
Give me a ping in case I can help testing.

File: generators/server/templates/src/main/kotlin/package/service/util/RandomUtil.kt.ejs [hunk 3]
```diff
@@ -20,44 +20,51 @@
...
+import java.security.SecureRandom
 import org.apache.commons.lang3.RandomStringUtils
 
 private const val DEF_COUNT = 20
 
+private val secureRandom: SecureRandom = SecureRandom().apply{ nextBytes(ByteArray(64)) }
+
+private fun generateRandomAlphanumericString() =
+    return RandomStringUtils.random(DEF_COUNT, 0, 0, true, true, null, secureRandom)
+}
```
Comment (sendilkumarn) [inline generators/server/templates/src/main/kotlin/package/service/util/RandomUtil.kt.ejs]:
```suggestion

```

File: generators/server/templates/src/main/kotlin/package/service/util/RandomUtil.kt.ejs [hunk 4]
```diff
@@ -20,44 +20,51 @@
...
- * Generate an activation key.
- *
- * @return the generated activation key.
- */
-fun generateActivationKey(): String = RandomStringUtils.randomNumeric(DEF_COUNT)
+* Generate an activation key.
+*
+* @return the generated activation key.
+*/
+fun generateActivationKey(): String = generateRandomAlphanumericString()
```
Comment (sendilkumarn) [inline generators/server/templates/src/main/kotlin/package/service/util/RandomUtil.kt.ejs]:
```suggestion
fun generateActivationKey() = generateRandomAlphanumericString()
```

Also for the others.

**Comment by atomfrede** [comment]
I got `Nothing' return type needs to be specified explicitly` in `RandomUtil.kt

**Comment by atomfrede** [comment]
Chrome can't be started. I have tested it with latest jhipster version and it worked fine locally. Should we merge this regardless of the failing tests?

**Comment by sendilkumarn** [comment]
@atomfrede Sorry for taking too long to reply. It is due to 6.3's mismatch in dependency. Can you base this PR against `move-to-6.3` branch. I will do a release once we completely migrate to 6.3.

**Comment by atomfrede** [comment]
@sendilkumarn Sure will do. Give me until tomorrow.

**Comment by atomfrede** [comment]
Done and rebased.

**Pull request merged**